In [67]:
import math
import numpy as np
import pandas as pd
from scipy import stats
from scipy import integrate

Задание №1. Рассчитайте следующие статистики для данных Task 1.csv (округлите до 2х знаков после запятой):

- Среднее;
- Стандартное отклонение;
- Дисперсию;
- Квартили (1-й, 2-й, 3-й, 4-й);
- Моду;
- 2-й перцентиль;
- Интерквартильный размах.

Примените те формулы, которые наиболее хорошо подходят для конкретного случая.

In [87]:
task1 = pd.read_csv('data/Biostatistics_HW1_Task_1.csv')

print('Среднее:', round(task1.mean()[0], 2))
print('Стандартное отклонение%', round(task1.std()[0], 2))
print('Дисперсия:', round(task1.var()[0], 2))

quantiles = task1.quantile([1/4, 1/2, 3/4, 1])

print('Квартили (1-й, 2-й, 3-й, 4-й):\n', round(quantiles, 2))
print('Мода:', round(stats.mode(task1).mode[0][0], 2))
print('2-й перцентиль:', round(np.percentile(task1, 2), 2))

interquartile_range = quantiles.iloc[3, 0] - quantiles.iloc[1, 0]
print('Интерквартильный размах:', round(interquartile_range, 2))



Среднее: 177.34
Стандартное отклонение% 115.26
Дисперсия: 13285.81
Квартили (1-й, 2-й, 3-й, 4-й):
       Values
0.25   99.23
0.50  151.07
0.75  220.37
1.00  928.66
Мода: 81.38
2-й перцентиль: 38.94
Интерквартильный размах: 777.59


/var/folders/10/w53fwtfx0136pc3wg8gmr_qh0000gn/T/ipykernel_18414/1941009269.py:10: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  print('Мода:', round(stats.mode(task1).mode[0][0], 2))


Задание № 2.1. Рассчитайте следующие статистики для данных Task 2_1.csv (округлите до 2х знаков после запятой):

- Среднее;
- Стандартное отклонение;
- Дисперсию;
- Квартили (1-й, 3-й).

Примените те формулы, которые наиболее хорошо подходят для конкретного случая.

In [89]:
task2_1 = pd.read_csv('data/Biostatistics_HW1_Task_2_1.csv')

print('Среднее:', round(task1.mean()[0], 2))
print('Стандартное отклонение:', round(task1.std()[0], 2))
print('Дисперсия:', round(task1.var()[0], 2))

quantiles = task1.quantile([1/4, 3/4])
print('Квартили (1-й, 3-й):\n', round(quantiles, 2))

Среднее: 177.34
Стандартное отклонение: 115.26
Дисперсия: 13285.81
Квартили (1-й, 3-й):
       Values
0.25   99.23
0.75  220.37


Задание №2.2. Рассчитайте следующие статистики для данных Task 2_2.csv (округлите до 2х знаков после запятой):

- Абсолютную частоту встречаемости каждого признака;
- Относительную частоту встречаемости каждого признака;
- Какой из признаков самый редко встречающийся.

In [91]:
task2_2 = pd.read_csv('data/Biostatistics_HW1_Task_2_2.csv')

counts = task2_2.value_counts()
print('Абсолютная частота встречаемости каждого признака:\n', counts)
counts_normalized = task2_2.value_counts(normalize = True)
print('Относительная частота встречаемости каждого признака:\n', round(counts_normalized, 2))

counts = counts.reset_index(name='Counts')
rare_car = counts[counts.Counts == counts.Counts.min()].Value.values[0]
print('Самый редко встречающийся признак:', rare_car)

Абсолютная частота встречаемости каждого признака:
 Value        
Purple Audi      30
Red Mercedes     27
Blue BMW         23
Jellow Toyota    22
dtype: int64
Относительная частота встречаемости каждого признака:
 Value        
Purple Audi      0.29
Red Mercedes     0.26
Blue BMW         0.23
Jellow Toyota    0.22
dtype: float64
Самый редко встречающийся признак: Jellow Toyota


Задание №3. Постройте доверительный 90% доверительный интервал для данных Task 2_1.csv.

Примечание: данные имеют распределение, отличное от нормального.

In [76]:
stats.t.interval(alpha=0.90, df=len(task2_1)-1, loc=np.mean(task2_1), scale=stats.sem(task2_1)) 

m = np.mean(task2_1.Values)
s = np.std(task2_1.Values)
interval = stats.t.interval(alpha=0.9, df=len(task2_1)-1, loc=m, scale=s/np.sqrt(len(task2_1)-1))

print('Доверительный 90% интервал:', interval)

Доверительный 90% интервал: (750.5649300131288, 1004.7886699868714)


/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/numpy/core/fromnumeric.py:3462: FutureWarning: In a future version, DataFrame.mean(axis=None) will return a scalar mean over the entire DataFrame. To retain the old behavior, use 'frame.mean(axis=0)' or just 'frame.mean()'
  return mean(axis=axis, dtype=dtype, out=out, **kwargs)
/var/folders/10/w53fwtfx0136pc3wg8gmr_qh0000gn/T/ipykernel_18414/718464657.py:1: DeprecationWarning: Use of keyword argument 'alpha' for method 'interval' is deprecated and wil be removed in SciPy 1.11.0. Use first positional argument or keyword argument 'confidence' instead.
  stats.t.interval(alpha=0.90, df=len(task2_1)-1, loc=np.mean(task2_1), scale=stats.sem(task2_1))
/var/folders/10/w53fwtfx0136pc3wg8gmr_qh0000gn/T/ipykernel_18414/718464657.py:5: DeprecationWarning: Use of keyword argument 'alpha' for method 'interval' is deprecated and wil be removed in SciPy 1.11.0. Use first positional argument or keyword argument 'confidenc

Задание №4. В задаче № 3 Вы построили 90% доверительный интервал. Скажите, принадлежит ли эта выборка генеральной совокупности со средним = 783.21? Почему? Ответьте развернуто.

Да, эта выборка принаждлежит генеральной совокупности, так как среднее входит в доверительный интервал (751, 1005).

Задание №5. Рассчитайте площадь под кривой на отрезке от -40 до +35 следующей функции:


В нашем случае среднее = 32, а стандартное отклонение = 7.5. Ответ округлите до 3 знаков после запятой.

Примечания:

Эта функция написана вручную, скорее всего Вы не найдёте её ни в одном из пакетов.

C интегрированием отлично справляется библиотека scipy;

In [74]:
def function(x):
    sigma = 7.5
    mu = 32
    return (1 / (sigma * math.sqrt(2 * math.pi))) * (math.e ** ((x - mu)**2 / (-2 * sigma**2))) * ((x - mu)**2 / (2 * sigma**2))

result = integrate.quad(function, -40, 35)

print('Площадь под кривой на отрезке от -40 до 35:', round(result[0], 3))

Площадь под кривой на отрезке от -40 до 35: 0.254
